src/jdk.jartool/share/classes/sun/tools/jar/FingerPrint.java

```diff
@@ -232,85 +243,73 @@ public int hashCode() {
-                Set<String> exceptionSet = new HashSet<>();
-                if (exceptions != null) {
-                    for (String e : exceptions) {
-                        exceptionSet.add(e);
+        public void accept(ClassElement cle) {
+            switch (cle) {
+                case InnerClassesAttribute ica -> {
+                    for (var icm : ica.classes()) {
+                        if (this.nestedClass && icm.outerClass().isPresent()
+                                && this.name.equals(icm.innerClass().asInternalName())
+                                && this.outerClassName == null) {
+                            this.outerClassName = icm.outerClass().get().asInternalName();
+                        }
                     }
                 }
-                // treat type descriptor as a proxy for signature because signature
-                // is usually null, need to strip off the return type though
-                int n;
-                if (desc != null && (n = desc.lastIndexOf(')')) != -1) {
-                    desc = desc.substring(0, n + 1);
-                    methods.add(new Method(access, name, desc, exceptionSet));
+                case FieldModel fm -> {
+                    if (isPublic(fm.flags())) {
+                        fields.add(new Field(fm.flags().flagsMask(),
+                                fm.fieldName().stringValue(),
+                                fm.fieldType().stringValue()));
+                    }
+                }
+                case MethodModel mm -> {
+                    if (isPublic(mm.flags())) {
+                        Set<String> exceptionSet = new HashSet<>();
+                        mm.findAttribute(Attributes.EXCEPTIONS).ifPresent(ea ->
+                                ea.exceptions().forEach(e ->
+                                        exceptionSet.add(e.asInternalName())));
+                        // treat type descriptor as a proxy for signature because signature
+                        // is usually null, need to strip off the return type though
+                        int n;
+                        var desc = mm.methodType().stringValue();
+                        if (desc != null && (n = desc.lastIndexOf(')')) != -1) {
+                            desc = desc.substring(0, n + 1);
+                            methods.add(new Method(mm.flags().flagsMask(),
+                                    mm.methodName().stringValue(), desc, exceptionSet));
+                        }
+                    }
                 }
+                case EnclosingMethodAttribute ema -> {
+                    if (this.nestedClass) {
+                        this.outerClassName = ema.enclosingClass().asInternalName();
+                    }
+                }
+                default -> {}
             }
-            return null;
         }
 
-        @Override
-        public void visitEnd() {
-            this.nestedClass = this.outerClassName != null;
```

Comment:
`this.nestedClass` is updated after the attributes are visited.   Suggest to rename `nestedClass` to `maybeNestedClass` and change `isNestedClass` to:
```
    public boolean isNestedClass() {
        return attrs.maybeNestedClass && attrs.outerClassName != null;
    }
```

Reply:
I've changed `nestedClass` according to the suggestion, thanks.